test/variable-subsetter/utilities.py

```diff
@@ -0,0 +1,123 @@
+""" A module containing utility functionality used by the Variable Subsetter
+    regression tests. These functions are kept out of the Jupyter notebook to
+    increase the readability of the regression test suite.
+
+"""
+from os import listdir, remove, replace
+from typing import Union
+
+from harmony import Client, Request
+from harmony.harmony import ProcessingFailedException
+from netCDF4 import Dataset, Group, Variable
+import numpy as np
+
+
+GroupOrVariable = Union[Group, Variable]
+
+
+def compare_attributes_to_reference(results_object: GroupOrVariable,
+                                    ref_object: GroupOrVariable):
+    """ Ensure the metadata attributes of two NetCDF-4 objects (groups or
+        variables) are the same, with the exception of `history` and
+        `history_json`, which will include references to the request time.
+
+    """
+    assert results_object.ncattrs() == ref_object.ncattrs()
+
+    for attribute_name, ref_attribute_value in ref_object.__dict__.items():
+        if attribute_name not in ['history', 'history_json']:
+            if isinstance(ref_attribute_value, np.ndarray):
+                np.testing.assert_array_equal(
+                    results_object.getncattr(attribute_name),
+                    ref_attribute_value
+                )
+            else:
+                assert results_object.getncattr(attribute_name) == ref_attribute_value
+
+
+def compare_variable_to_reference(results_variable: Variable,
+                                  ref_variable: Variable):
+    """ Compare two NetCDF-4 variables, ensuring they have the same data in
+        their arrays and the same metadata attribute.
+
+    """
+    np.testing.assert_array_equal(results_variable[:], ref_variable[:])
+    compare_attributes_to_reference(results_variable, ref_variable)
+
+
+def compare_group_to_reference(results_group: Group, ref_group: Group):
+    """ Compare two NetCDF-4 file groups, ensuring they have the same metadata
+        attributes (excluding provenance), child variables and child groups.
+        Child variables and groups are then compared recursively.
+
+    """
+    compare_attributes_to_reference(results_group, ref_group)
+
+    assert list(results_group.groups.keys()) == list(ref_group.groups.keys())
+    assert list(results_group.variables.keys()) == list(ref_group.variables.keys())
+
+    for variable_name, ref_variable in ref_group.variables.items():
+        compare_variable_to_reference(results_group[variable_name], ref_variable)
+
+    for child_group_name, ref_child_group in ref_group.groups.items():
+        compare_group_to_reference(results_group[child_group_name],
+                                   ref_child_group)
+
+
+def compare_results_to_reference_file(results_file: str, ref_file: str):
```

Comment:
This version is beefier than the `xarray` method because `xarray` doesn't handle hierarchies in NetCDF-4 or HDF-5 files very well. But I used that where possible to keep the amount of code down.